docs/writing-controllers.md

```diff
@@ -0,0 +1,1489 @@
+    super({ messenger /*, ... */ });
+  }
+
+  doSomething() {
+    this.messagingSystem.publish('FooController:someEvent');
+  }
+}
+
+// === Client repo ===
+
+const rootMessenger = new ControllerMessenger<
+  'FooController:someEvent',
+  never
+>();
+const fooControllerMessenger = rootMessenger.getRestricted({
+  name: 'FooController',
+});
+const fooController = new FooController({
+  messenger: fooControllerMessenger,
+});
+rootMessenger.subscribe('FooController:someEvent', () => {
+  // do something with the event
+});
+```
+
+## Define the `*:getState` action using the `ControllerGetStateAction` utility type
+
+Each controller needs a type for its `*:getState` action. The `ControllerGetStateAction` utility type from the `@metamask/base-controller` package should be used to define this type.
+
+The name of this type should be `${ControllerName}GetStateAction`.
+
+```typescript
+import type { ControllerGetStateAction } from '@metamask/base-controller';
+
+type FooControllerGetStateAction = ControllerGetStateAction<
+  'FooController',
+  FooControllerState
+>;
+```
+
+## Define the `*:stateChange` event using the `ControllerStateChangeEvent` utility type
+
+Each controller needs a type for its `*:stateChange` event. The `ControllerStateChangeEvent` utility type from the `@metamask/base-controller` package should be used to define this type.
+
+The name of this type should be `${ControllerName}StateChangeEvent`.
+
+```typescript
+import type { ControllerStateChangeEvent } from '@metamask/base-controller';
+
+type FooControllerStateChangeEvent = ControllerStateChangeEvent<
+  'FooController',
+  FooControllerState
+>;
+```
+
+## Use standard naming scheme for action identifiers
+
+An action is equivalent to a function, and its identifier should be named in "camelCase".
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerShowNotificationAction = {
+  type: `${typeof name}:showNotification`;
+  handler: () => void;
+};
+```
+
+## Use standard naming scheme for action types
+
+Types for messenger actions should follow this naming scheme:
+
+```
+Controller name + Action identifier without controller name prefix + "Action"
+```
+
+🚫 **Type name does not include controller name and does not end with "Action"**
+
+```typescript
+const name = 'FooController';
+
+export type ShowNotification = {
+  type: `${typeof name}:show`;
+  handler: () => void;
+};
+```
+
+🚫 **Type name includes controller name, but does not match action identifier minus prefix**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerShowNotificationAction = {
+  type: `${typeof name}:show`;
+  handler: () => void;
+};
+```
+
+✅ **Type name now matches controller name + action identifier + "Action"**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerShowAction = {
+  type: `${typeof name}:show`;
+  handler: () => void;
+};
+```
+
+## Use standard naming for event identifiers
```

Comment:
For consistency with https://github.com/MetaMask/core/pull/4110#discussion_r1670855179

```suggestion
## Use standard naming scheme for event identifiers
```